### PR TITLE
ENH: add ndmin to `genfromtxt` behaving the same as `loadtxt`

### LINKED ISSUE #4811: Enhancement request: add ndmin option to genfromtxt
A couple years ago, `loadtxt` was given the argument `ndmin`.   `genfromtxt` should also have this option.

See, for example, http://stackoverflow.com/questions/24247103/numpy-genfromtxt-collapse-recarray-when-data-file-has-only-one-row/

### PR DESCRIPTION
**Pull request by scratchmex**
A workaround to avoid this error https://github.com/numpy/numpy/issues/20451

I also add the proper types annotations

closes https://github.com/numpy/numpy/issues/4811 a very old ENH request

Feel free to comment.

Should we make `ndmin=1` as default parameter to avoid this https://github.com/numpy/numpy/issues/20494?

### CODE REVIEW

**Comment by scratchmex** [comment]
@InessaPawson here is some contribution to close very old stuff :)

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -1687,6 +1708,8 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
         to None the system default is used. The default value is 'bytes'.
 
         .. versionadded:: 1.14.0
+    ndmin : int, optional
+        Same parameter as `loadtxt`
```
Comment (BvB93) [inline numpy/lib/npyio.py:1712]: This will require a `versionadded` note.

File: numpy/lib/npyio.py [hunk 2]
```diff
@@ -1573,8 +1593,9 @@ def _genfromtxt_dispatcher(fname, dtype=None, comments=None, delimiter=None,
                            names=None, excludelist=None, deletechars=None,
                            replace_space=None, autostrip=None, case_sensitive=None,
                            defaultfmt=None, unpack=None, usemask=None, loose=None,
-                           invalid_raise=None, max_rows=None, encoding=None, *,
-                           like=None):
+                           invalid_raise=None, max_rows=None, encoding=None,
+                           ndmin=None,
+                           *, like=None):
```
Comment (BvB93) [inline numpy/lib/npyio.py:1598]:
```suggestion
                           invalid_raise=None, max_rows=None, encoding=None, *,
                           ndmin=None, like=None):
```
New optional arguments are usually made keyword-only these days.

**Review by BvB93 — COMMENTED** [review]
I've got a few comments, mostly about the added type annotations.
There are two more things of note here:
* The comments apply to all overloads.
* The argument order in the .pyi stubfile does not match its .py counterpart. This should be fixed.

File: numpy/lib/npyio.py (hunk 1 above)
Comment (scratchmex) [reply numpy/lib/npyio.py:1712]: which version should I put?

File: numpy/lib/npyio.pyi [hunk 3]
```diff
@@ -212,27 +212,86 @@ def fromregex(
...
+    usecols: Sequence[int] = ...,
+    names: Union[None, True, str, Sequence] = ...,
+    excludelist: Sequence[str] = ...,
+    deletechars: str = ...,
+    replace_space: str = ...,
+    autostrip: bool = ...,
+    case_sensitive: L[True, False, 'upper', 'lower'] = ...,
+    defaultfmt: str = ...,
+    unpack: bool = ...,
+    usemask: bool = ...,
```
[the remark this answers is not in the document]
Comment (scratchmex) [reply numpy/lib/npyio.pyi:262]: `usemask` default value is `False` if a I'm not mistaken

[the remark this answers is not in the document]
Comment (scratchmex) [reply numpy/lib/npyio.pyi:262]: I think you are refering to `unpack=None`. I will fix that

File: numpy/lib/npyio.py (hunk 1 above)
Comment (scratchmex) [reply numpy/lib/npyio.py:1711]: I think it is `1.23`

File: numpy/lib/npyio.pyi (hunk 3 above)
[the remark this answers is not in the document]
Comment (BvB93) [reply numpy/lib/npyio.pyi:265]: Ah you're right, seem I marked the wrong parameter.

File: numpy/lib/npyio.py (hunk 1 above)
Comment (BvB93) [reply numpy/lib/npyio.py:1711]: I'd put it at the bottom of the `ndim` description. The `max_rows` docs also have a `versionadded` note, I'd use that as example if I were you.

**Comment by BvB93** [comment]
Ah hang on, I see I made a mistake earlier in one of the [typing tests](https://github.com/numpy/numpy/blob/0e10696f55576441fd820279d9ec10cd9f2a4c5d/numpy/typing/tests/data/reveal/npyio.pyi#L79).
Removing the `skiprows` parameter should fix the failing test:
``` diff
- reveal_type(np.genfromtxt(str_path, dtype=str, skiprows=2))  # E: ndarray[Any, dtype[Any]]
+ reveal_type(np.genfromtxt(str_path, dtype=str))  # E: ndarray[Any, dtype[Any]]
```

**Comment by scratchmex** [comment]
> ```diff
> reveal_type(np.genfromtxt(str_path, dtype=str))
> ```

Better, I replaced `skiprows` with `skip_header` and now it works :)

File: doc/release/upcoming_changes/20500.new_feature.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+New parameter `ndmin` added to `genfromtxt`
+-------------------------------------------------------------------------
+This parameter behaves the same as `ndmin` from `loadtxt`.
```
Comment (BvB93) [inline doc/release/upcoming_changes/20500.new_feature.rst:3]:
```suggestion
New parameter ``ndmin`` added to ``genfromtxt``
-------------------------------------------------------------------------
This parameter behaves the same as ``ndmin`` from `~numpy.loadtxt`.
```

File: numpy/lib/tests/test_io.py [hunk 5]
```diff
@@ -2438,6 +2438,16 @@ def test_squeeze_scalar(self):
         assert_equal((), test.shape)
         assert_equal(expected.dtype, test.dtype)
 
+    def test_ndmin_keyword(self):
+        # lets have the same behaivour of ndmin as loadtxt
+        # as they should be the same for non-missing values
+        txt = "42"
+
+        a = np.loadtxt(StringIO(txt), ndmin=1)
+        b = np.genfromtxt(StringIO(txt), ndmin=1)
```
Comment (BvB93) [inline numpy/lib/tests/test_io.py:2447]:
```suggestion
    @pytest.mark.parametrize("ndim", [0, 1, 2])
    def test_ndmin_keyword(self, ndim: int):
        # lets have the same behaivour of ndmin as loadtxt
        # as they should be the same for non-missing values
        txt = "42"

        a = np.loadtxt(StringIO(txt), ndmin=ndim)
        b = np.genfromtxt(StringIO(txt), ndmin=ndim)
```
I'd parametrize over all valid dimensions here; this should improve the test coverage a somewhat.

File: doc/release/upcoming_changes/20500.new_feature.rst (hunk 4 above)
Comment (BvB93) [reply doc/release/upcoming_changes/20500.new_feature.rst:3]: Making appropriate use of single and double quotes should help with intersphinx links.

**Review by BvB93 — COMMENTED** [review]
Looks good to me.
I've got two more minor comments but the PR should be good to go afterwards.

**Review by BvB93 — APPROVED** [review]

**Comment by BvB93** [comment]
In it goes, thanks @scratchmex.

**Comment by emirkmo** [comment]
Very cool to dig up that old issue and close it, thanks. Even 5 years later, this is a helpful change. Makes it easier to get rid of certain libraries that are used just to parse data into np data structures.

**Pull request merged**